File: api-gateway/src/main/kotlin/at/fhcampuswien/dev/we/resource/order/OrderService.kt
package at.fhcampuswien.dev.we.resource.order

import at.fhcampuswien.dev.we.messaging.OrderMessageProducerService
import at.fhcampuswien.dev.we.messaging.OrderRPCService
import at.fhcampuswien.dev.we.messaging.StationMessageService
import at.fhcampuswien.dev.we.messaging.StationRPCService
import at.fhcampuswien.dev.we.order.model.order.OrderBillingDTO
import at.fhcampuswien.dev.we.order.model.order.OrderDTO
import at.fhcampuswien.dev.we.order.model.station.StationOrderDto
import jakarta.inject.Singleton


@Singleton
class OrderService(
    private val messageProducerService: OrderMessageProducerService,
    private val stationRPCService: StationRPCService,
    private val orderRPCService: OrderRPCService,
    private val stationMessageService: StationMessageService
) {

    fun createOrder(order: OrderDTO): OrderDTO {
        messageProducerService.sendCreateCommand(order)
        return order
    }

    fun billOrder(deliverTo: String): String {
        messageProducerService.sendBillingCommand(deliverTo)
        return deliverTo
    }

    fun getOrders(): List<OrderDTO> {
        return orderRPCService.getOrders()
    }

    fun getOrdersForBilling(): List<OrderBillingDTO> {
        return orderRPCService.getOrdersForBilling()
    }

    fun getOrdersByStation(stationType: String): List<StationOrderDto> {
        return stationRPCService.getAllByStation(stationType)
    }

    fun finishStationOrder(stationOrderId: String) {
        stationMessageService.sendFinishCommand(stationOrderId)
    }
}

File: frontend/src/app/core/services/order.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { OrderDto } from '../model/order/order.dto';
import { Observable } from 'rxjs';
import { StationOrderDto } from '../model/station/station-order.dto';
import { environment } from '../../../environments/environment';
import { OrderBillingDto } from '../model/order/oder-billing.dto';

@Injectable({
  providedIn: 'root'
})
export class OrderService {

  constructor(private httpClient: HttpClient) { }

  createOrder(order: OrderDto): Observable<any> {
    return this.httpClient.post<OrderDto>(environment.apiBaseUrl + '/order/create', order);
  }

  getOrders(): Observable<OrderDto[]> {
    return this.httpClient.get<OrderDto[]>(environment.apiBaseUrl + '/order');
  }

  getOrdersForBilling(): Observable<OrderBillingDto[]> {
    return this.httpClient.get<OrderBillingDto[]>(environment.apiBaseUrl + '/order/billing');
  }

  billOrder(deliverTo: string): Observable<any> {
    return this.httpClient.post(environment.apiBaseUrl + '/order/bill', deliverTo, {responseType: 'text'});
  }

  getOrdersByStation(stationType: string): Observable<StationOrderDto[]> {
    return this.httpClient.get<StationOrderDto[]>(environment.apiBaseUrl + '/order/station/' + stationType);
  }

  finishStationOrder(stationOrderId: string): Observable<any> {
    return this.httpClient.post(environment.apiBaseUrl + '/order/station/finish', stationOrderId, {responseType: 'text'});
  }
}
